File: Seki/seki_tester.py

from direct_solver import SekiSolver
from direct_solver_with_get_moves import SekiSolver as SekiSolver_wgm
from numpy import array
import sys
import time
height, width = map(int, input("height, width:").split())

raw = []
print("input matrix:")
for i in range(height):
  arr = [*map(int, input().split())]
  if len(arr) != width:
    print("Row too long!", file=sys.stderr)
    sys.exit(1)
  raw.append(arr)

field = array(raw, dtype=int)
solver = SekiSolver(field, type="seki")

while True:
  print(solver.field.data)
  p = (*map(int, input("your move(x, y):").split()),)
  solver.decrement(*p)
  start = time.time() # start counting time
  pr = solver.find_optimal(False)
  end = time.time() # end counting time
  if pr is None:
    print("No optimal responce for C")
    break
  print('time:', end - start,'|matrices per second:', solver.unrolled/(end - start), '|pr.value:', pr.value, '|pr.x:', pr.x, '|pr.y:', pr.y)
  solver.decrement(pr.x, pr.y)
  if solver.field.is_terminal():
    print('game finished\n', solver.field.data)
    break

File: Seki/seki_tester.cpp
#include "direct_solver.cpp"
#include <iostream>
#include <random>
#include <chrono>

int main() {
  srand(time(NULL));
  int n, m;
  cout << "height, width\n";
  cin >> n >> m;
  vector<vector<int>> v;
  for (int i = 0; i < n; ++i) {
    v.push_back(vector<int>());
    for (int j = 0; j < m; ++j) {
      int f;
      cin >> f;
      v.back().push_back(f);
    }
  }

  auto ss = SekiSolver(v, SekiType::SEKI);
  while (!ss.state.is_terminal()) {
    int x, y;
    cout << ss.state;
    cout << "x,y: ";
    cin >> x >> y;
    ss.decrement(x, y);
    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    Move opt = ss.find_optimal(false);
    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    double time_seconds = std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count() / 1000000.0;
    cout << "time: " << time_seconds << "|matrices per second: " << ss.unrolled / time_seconds  <<  "|pr.value: " << opt.value << "|pr.x " << opt.x << "|pr.y " << opt.y << endl;
    ss.decrement(opt.x, opt.y);
  }

  cout << ss.state;

  return 0;
}